integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala

```diff
@@ -148,10 +149,19 @@ case class CarbonInsertFromStageCommand(
         return Seq.empty
       }
 
-      // 2) read all stage files to collect input files for data loading
-      // create a thread pool to read them
+      // We add a tag 'loading' to the stages in process.
+      // different insertstage processes can load different data separately
+      // by choose the stages without 'loading' tag or stages loaded timeout.
+      // which avoid loading the same data between concurrent insertstage processes.
+      // The 'loading' tag is actually an empty file with
+      // '.loading' suffix filename
       val numThreads = Math.min(Math.max(stageFiles.length, 1), 10)
       val executorService = Executors.newFixedThreadPool(numThreads)
+      createStageLoadingFilesWithRetry(executorService, stageFiles)
+      lock.unlock()
```

Comment:
remove this line, lock will unlock in finally block.

Reply:
it can't be removed, as we aim to release ingest lock once complete tag 'loading' for the choosed stage.